### PR TITLE
fix NumberFormatException for invalid noise values

### LINKED ISSUE #1366: Follower receives NS entries download Failed: NumberFormatException - Received Null
Hi,

I have set up my NightScout account, and xDrip can write to it from the master device, but when setting up a following device, I receive the following -

NS entries download Failed:
com.google.gson.JsonSyntaxException:
java.lang.NumberFormatException: For input string: "Null"

The NightScout account was originally populated with entries by Diabox, but this stopped after a few hours, and I can't find a way of getting the details on a Wear OS watch. Hence why I'm trying xDrip. So there are gaps in the data, but I wouldn't expect that to cause an issue.

Thanks,

Brian

### PR DESCRIPTION
**Pull request by tolot27**
Attempt to fix #1366 for new and existing noise values.

### CODE REVIEW

File: app/src/main/java/com/eveningoutpost/dexdrip/cgm/nsfollow/NightscoutFollow.java [hunk 1]
```diff
@@ -150,10 +194,13 @@ public static Retrofit getRetrofitInstance() throws IllegalArgumentException {
                     .addInterceptor(new GzipRequestInterceptor())
                     .build();
 
+            final Gson gson = new GsonBuilder()
+                    .registerTypeAdapterFactory(UNRELIABLE_INTEGER_FACTORY)
+                    .create();
```
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/cgm/nsfollow/NightscoutFollow.java:199]: We may need to add `.serializeNulls()`.

**Comment by tolot27** [comment]
I'll also add a test case.

File: app/src/main/java/com/eveningoutpost/dexdrip/cgm/nsfollow/NightscoutFollow.java (hunk 1 above)
Comment (tolot27) [reply app/src/main/java/com/eveningoutpost/dexdrip/cgm/nsfollow/NightscoutFollow.java:199]: `serializeNulls` has an effect on serialization only, not during deserialization. Hence, not needed at that code point.

**Comment by tolot27** [comment]
PR is ready for review.

**Comment by tolot27** [comment]
It looks like the source of the problem was fixed by the diabox developers. Nevertheless, this patch is still useful in case of improperly formatted numbers in json strings from other sources.

**Comment by jamorham** [comment]
Can you briefly describe what the code changes do? I only looked briefly but it looks that you are adding a more tolerant parser for integers to the json decoding?

**Comment by tolot27** [comment]
> Can you briefly describe what the code changes do? I only looked briefly but it looks that you are adding a more tolerant parser for integers to the json decoding?

Yes, this PR adds a special TypeAdapter for native ints and Integer objects which gets called during deserialization of a json stream. Within this TypeAdapter, several string to number convert exceptions are catched and a default value is set.
There are some type differences/inconsistencies between the `Entry` class and the `BgReading` class, for example, noise is declared as `int` in `Entry` whereas it is declared as `String` in `BgReading`. That could lead to different json representations which can cause problems during deserialization. This PR tries to fix this during serialization and deserialization without touching the declarations.

**Review by jamorham — APPROVED** [review]
This looks good to me. I assume that there isn't an issue where something intended as a string which contains something that can be parsed as a number gets misrepresented as a result of these changes.

**Pull request merged**